pennylane/shadows/classical_shadow.py

```diff
@@ -0,0 +1,370 @@
+        This corresponds to the case ``k=1``. However, it is often desirable for better accuracy to split the ``T`` measurements into ``k`` equal parts to compute the median of means, see `2002.08953 <https://arxiv.org/abs/2002.08953>`_.
+
+        One of the main perks of classical shadows is being able to compute many different expectation values by classically post-processing the same measurements. This is helpful in general as it may help
+        save quantum circuit executions.
+
+        Args:
+            H (:class:`~.pennylane.Hamiltonian` or :class:`~.pennylane.operation.Tensor`): Observable to compute the expectation value over.
+            k (int): Number of equal parts to split the shadow's measurements to compute the median of means. ``k=1`` corresponds to simply taking the mean over all measurements.
+
+        Args:
+            H (qml.Observable): Observable to compute the expectation value
+            k (int): Split the snapshots into ``k`` equal parts to compute the median of means.
+
+        Returns:
+            float: expectation value estimate.
+
+        **Example**
+
+        .. code-block:: python3
+
+            dev = qml.device("default.qubit", wires=range(2), shots=1000)
+            @qml.qnode(dev)
+            def qnode(x):
+                qml.Hadamard(0)
+                qml.CNOT((0,1))
+                qml.RX(x, wires=0)
+                return classical_shadow(wires=range(2))
+
+            bits, recipes = qnode(0)
+            shadow = ClassicalShadow(bits, recipes)
+
+        Compute Pauli string observables
+
+        >>> shadow.expval(qml.PauliX(0) @ qml.PauliX(1), k=1)
+        (1.0079999999999998+0j)
+
+        or of a Hamiltonian using `the same` measurement results
+
+        >>> H = qml.Hamiltonian([1., 1.], [qml.PauliZ(0)@qml.PauliZ(1), qml.PauliX(0)@qml.PauliX(1)])
+        >>> shadow.expval(H, k=1)
+        (2.2319999999999998+0j)
+        """
+        coeffs_and_words = self._convert_to_pauli_words(H)
+
+        expval = 0
+        for coeff, word in coeffs_and_words:
+            expvals = pauli_expval(self.bits, self.recipes, np.array(word))
+            expval += coeff * median_of_means(expvals, k)
+
+        return expval
+
+
+# Util functions
+def median_of_means(arr, num_batches):
+    r"""
+    The median of means of the given array.
+
+    The array is split into the specified number of batches. The mean value
+    of each batch is taken, then the median of the mean values is returned.
+
+    Args:
+        arr (tensor-like[float]): The 1-D array for which the median of means
+            is determined
+        num_batches (int): The number of batches to split the array into
+
+    Returns:
+        float: The median of means
+    """
+    means = []
+    batch_size = int(np.ceil(arr.shape[0] / num_batches))
+
+    for i in range(num_batches):
+        means.append(qml.math.mean(arr[i * batch_size : (i + 1) * batch_size], 0))
```

Comment:
Better to use list comprehension.
```suggestion
    means = [qml.math.mean(arr[i * batch_size : (i + 1) * batch_size], 0) for i in range(num_batches)]
```